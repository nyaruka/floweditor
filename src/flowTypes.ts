import {
    IsOptionUniqueHandler,
    IsValidNewOptionHandler,
    NewOptionCreatorHandler,
    PromptTextCreatorHandler
} from 'react-select';

export interface Languages {
    [iso: string]: string;
}

export interface Endpoints {
    fields: string;
    groups: string;
    engine: string;
    contacts: string;
    flows: string;
    activity: string;
}

export interface FlowEditorConfig {
    languages: { [iso: string]: string };
    endpoints: Endpoints;
    flow: string;
    path?: string;
}

export interface LocalizationMap {
    [lang: string]: {
        [uuid: string]: any;
    };
}

export interface FlowDefinition {
    localization: LocalizationMap;
    language: string;
    name: string;
    nodes: FlowNode[];
    uuid: string;
    _ui: UIMetaData;
}

export interface FlowNode {
    uuid: string;
    actions: Action[];
    exits: Exit[];
    router?: Router;
    wait?: any;
}

export interface Exit {
    uuid: string;
    name?: string;
    destination_node_uuid?: string;
}

export interface Router {
    type: string;
    result_name?: string;
}

export interface Case {
    uuid: string;
    type: string;
    exit_uuid: string;
    arguments?: string[];
}

export interface SwitchRouter extends Router {
    cases: Case[];
    operand: string;
    default_exit_uuid: string;
}

export enum WaitType {
    exp = 'exp',
    group = 'group',
    msg = 'msg',
    flow = 'flow'
}

export interface Wait {
    type: WaitType;
    flow_uuid?: string;
}

export interface Group {
    uuid: string;
    name: string;
}

export enum Methods {
    GET = 'GET',
    POST = 'POST',
    PUT = 'PUT'
}

export interface Action {
    type: string;
    uuid: string;
}

export interface ChangeGroups extends Action {
    groups: Group[];
}

export interface Field {
    key: string;
    name: string;
}

export interface SetContactField extends Action {
    field: Field;
    value: string;
}

export interface SetContactProperty extends Action {
    property: string;
    value: string;
}

export type SetContactAttribute = SetContactField | SetContactProperty;

export interface SendMsg extends Action {
    text: string;
    all_urns?: boolean;
}

export interface SetPreferredChannel extends Action {
    language: string;
}

export interface SendEmail extends Action {
    subject: string;
    body: string;
    emails: string[];
}

export interface SetRunResult extends Action {
    result_name: string;
    value: string;
    category?: string;
}

export interface Headers {
    [name: string]: string;
}

export interface CallWebhook extends Action {
    url: string;
    method: Methods;
    body?: string;
    headers?: Headers;
}

export interface StartFlow extends Action {
    flow_name: string;
    flow_uuid: string;
}

export interface UIMetaData {
    nodes: { [key: string]: UINode };
    languages: Array<{ [iso: string]: string }>;
}

export interface FlowPosition {
    left: number;
    top: number;
    right?: number;
    bottom?: number;
}

export interface Dimensions {
    width: number;
    height: number;
}

export interface UINode {
    position: FlowPosition;

    // ui type, used for split by expression, contact field, etc
    type?: string;
}

export interface StickyNote {
    position: FlowPosition;
    title: string;
    body: string;
    color?: string;
}

export interface UIMetaData {
    nodes: { [key: string]: UINode };
    stickies: { [key: string]: StickyNote };
}

export type AnyAction =
    | Action
    | ChangeGroups
    | SetContactField
    | SetContactProperty
    | SetRunResult
    | SendMsg
    | SetPreferredChannel
    | SendEmail
    | CallWebhook
    | StartFlow;

export enum ContactProperties {
    UUID = 'UUID',
    'Created By' = 'Created By',
    'Modified By' = 'Modified By',
    Org = 'Org',
    Name = 'Name',
    Language = 'Language',
    Timezone = 'Timezone',
    Email = 'Email',
    Mailto = 'Mailto',
    Phone = 'Phone',
    Groups = 'Groups',
    Facebook = 'Facebook',
    Telegram = 'Telegram'
}

export enum ResultType {
    flow = 'flow',
    field = 'field',
    group = 'group',
    label = 'label'
}

export enum ValueType {
    text = 'text',
    numeric = 'numeric',
    datetime = 'datetime',
    state = 'state',
    district = 'district',
    ward = 'ward'
}

export enum AttributeType {
    property = 'property',
    field = 'field'
}

export interface CreateOptions {
    promptTextCreator?: PromptTextCreatorHandler;
    newOptionCreator?: NewOptionCreatorHandler;
    isValidNewOption?: IsValidNewOptionHandler;
    isOptionUnique?: IsOptionUniqueHandler;
    createNewOption?: NewOptionCreatorHandler;
    createPrompt?: string;
}
