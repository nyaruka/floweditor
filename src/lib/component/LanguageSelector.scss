@import "../../variables.scss";

.language_selector {
    z-index: $z_editor_menu;
    position: absolute;
    right: 40px;
    top: 20px;
    width: 180px;
}
