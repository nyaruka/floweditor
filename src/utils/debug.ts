import { DebugState } from 'store/editor';
import { store } from 'store';

const mutate = require('immutability-helper');

/* istanbul ignore next */
export default class Debug {
  private state: DebugState;

  constructor(props: any, initial: DebugState) {
    this.state = initial || { showUUIDs: false };
  }

  public showUUIDs(): DebugState {
    const updated = mutate(this.state, { $merge: { showUUIDs: true } });
    
    // Update via TembaStore instead of Redux
    const currentEditorState = store.getState().editorState;
    store.getState().updateEditorState({ 
      ...currentEditorState, 
      debug: updated 
    });
    
    return updated;
  }
}
