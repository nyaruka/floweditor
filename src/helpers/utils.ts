import { ComponentClass, SFC, ReactElement } from 'react';
import { ShallowWrapper, ReactWrapper } from 'enzyme';
import { FlowDefinition } from '../flowTypes';
import { CharacterSet } from '../components/form/TextInputElement';

const SNAKED_CHARS = /\s+(?=\S)/g;
const V4_UUID = /^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/i;

interface BoolMap {
    [key: string]: boolean;
}

/**
 * Turns a string array into a bool map for constant lookup
 * @param {string[]} array - an array of strings, e.g. contact fields
 * @returns {object} A map of each string
 */
export const toBoolMap = (array: string[]): BoolMap =>
    array.reduce(
        (map: BoolMap, item: string) => ({
            ...map,
            [item]: true
        }),
        {}
    );

/**
 * Inserts commas into numbers where appropriate for better readability
 * @param {number} value - A number, e.g. 10000
 * @returns {string} A comma-separated string, e.g. 10,000
 */
export const addCommas = (value: number): string =>
    value.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');

/**
 * Replaces spaces with underscores
 * @param {string} value - A space-separated string to be snaked, e.g. a raw flow field name ('my flow field')
 * @returns {string} A snaked string, e.g. 'my_flow_field'
 */
export const snakify = (value: string): string =>
    value
        .toLowerCase()
        .trim()
        .replace(SNAKED_CHARS, '_');

/**
 * Returns a React component's name so we can attach it to a HOC's displayName property and view it in RDT
 * @param {ComponentClass | SFC} Component - A React component
 * @returns {string} The component's name
 */
export const getDisplayName = (WrappedComponent: ComponentClass | SFC): string =>
    WrappedComponent.displayName || WrappedComponent.name || 'Component';

/**
 * NOTE: borrowed from EventBrite: https://github.com/eventbrite/javascript/blob/master/react/testing.md#finding-nodes
 * Finds all instances of components in the rendered `componentWrapper` that are DOM components
 * with the `data-spec` attribute matching `name`.
 * @param {ReactWrapper|ShallowWrapper} componentWrapper - Rendered componentWrapper (result of mount, shallow, or render)
 * @param  {string} snacName - Name of `data-spec` attribute value to find
 * @param {string|Function} (Optional) typeFilter - (Optional) Expected type of the wrappers (defaults to all HTML tags)
 * @returns {ReactWrapper|ReactWrapper[]|ShallowWrapper|ShallowWrapper[]} Matching DOM components
 */
export const getSpecWrapper = (
    componentWrapper: ReactWrapper | ShallowWrapper,
    specName: string,
    attributeName: string = 'data-spec'
): any => {
    return componentWrapper.find(`[${attributeName}="${specName}"]`);
};

/**
 * Returns true if a given UUID matches v4 format
 * @param {string} uuid - A version 4 UUID (no braces, uppercase OK)
 * @returns {boolean}
 */
export const validUUID = (uuid: string): boolean => V4_UUID.test(uuid);

/**
 * Returns a given string in title case, e.g. 'full name' becomes 'Full Name'
 * @param {string} str - string to be title-cased
 * @returns {string} Title-cased string
 */
export const titleCase = (str: string): string =>
    str.replace(/\b\w+/g, s => s.charAt(0).toUpperCase() + s.substr(1).toLowerCase());

export const getSelectClass = (errors: number): string => {
    if (errors === 0) {
        return '';
    }
    // We use a global selector here for react-select
    return 'select-invalid';
};

export const reorderList = (list: any[], startIndex: number, endIndex: number): any[] => {
    const [removed] = list.splice(startIndex, 1);

    list.splice(endIndex, 0, removed);

    return list;
};

/**
 * Compares basic objects (no methods and DOM nodes; property order important)
 * @param {object} objA - basic object
 * @param {object} objB - basic object
 * @returns {boolean}
 */
export const jsonEqual = (objA: {}, objB: {}): boolean =>
    JSON.stringify(objA) === JSON.stringify(objB);

/**
 * Checks whether any of a list of error strings contain one or more queries.
 * Used to determine whether a particular error has been encountered.
 * @param {Array.<string>} errors - list of error messages
 * @param {Array.<string>} queries  - list of queries
 * @returns {boolean}
 */
export const hasErrorType = (errors: string[], queries: string[]): boolean => {
    if (!errors.length) {
        return false;
    }

    for (const error of errors) {
        for (const query of queries) {
            if (error.indexOf(query) > -1) {
                return true;
            }
        }
    }

    return false;
};
