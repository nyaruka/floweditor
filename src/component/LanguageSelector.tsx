import * as React from 'react';
import { connect } from 'react-redux';
import Select from 'react-select';
import { bindActionCreators } from 'redux';
import { getBaseLanguage } from '.';
import { Config } from '../config';
import { Languages } from '../flowTypes';
import {
    DispatchWithState,
    ReduxState,
    UpdateLanguage,
    updateLanguage,
    UpdateTranslating,
    updateTranslating
} from '../redux';
import { jsonEqual } from '../utils';
import { languageSelector } from './LanguageSelector.scss';

export interface Language {
    name: string;
    iso: string;
}

export interface LanguageSelectorPassedProps {
    languages: Languages;
}

export interface LanguageSelectorDuxProps {
    language: Language;
    updateLanguage: UpdateLanguage;
    updateTranslating: UpdateTranslating;
}

export type LanguageSelectorProps = LanguageSelectorPassedProps & LanguageSelectorDuxProps;

export const composeLanguageMap = (languages: Languages): Language[] =>
    Object.keys(languages).map(iso => ({
        name: languages[iso],
        iso
    }));

const LanguageSelectorContainer: React.SFC = () => (
    <Config
        render={({ languages }) =>
            // prettier-ignore
            <ConnectedLanguageSelector
                languages={languages}
            />
        }
    />
);

const LanguageSelector: React.SFC<LanguageSelectorProps> = ({
    language,
    languages,
    updateLanguage,
    updateTranslating
}) => {
    if (language) {
        const onChange = (lang: Language) => {
            const baseLanguage = getBaseLanguage(languages);
            const translating = !jsonEqual(baseLanguage, lang);
            updateLanguage(lang);
            updateTranslating(translating);
        };

        const options = composeLanguageMap(languages);

        return (
            <div className={`${languageSelector} select-small`}>
                <Select
                    data-spec="language-selector"
                    // Flow
                    value={language.iso}
                    onChange={onChange}
                    // LanguageSelector
                    valueKey="iso"
                    labelKey="name"
                    searchable={false}
                    clearable={false}
                    options={options}
                />
            </div>
        );
    }

    return null;
};

const mapStateToProps = ({ language }: ReduxState) => ({ language });

const mapDispatchToProps = (dispatch: DispatchWithState) =>
    bindActionCreators(
        {
            updateLanguage,
            updateTranslating
        },
        dispatch
    );

const ConnectedLanguageSelector = connect(mapStateToProps, mapDispatchToProps)(LanguageSelector);

export default LanguageSelectorContainer;
