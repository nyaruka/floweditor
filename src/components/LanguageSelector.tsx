import * as React from 'react';
import Select from 'react-select';
import { Languages } from '../flowTypes';

import { languageSelector } from './LanguageSelector.scss';

export interface Language {
    name: string;
    iso: string;
}

export interface LanguageSelectorProps {
    iso: string;
    languages: Languages;
    onChange(language: Language): void;
}

export const composeLanguageMap = (languages: Languages): Language[] =>
    Object.keys(languages).map(iso => ({
        name: languages[iso],
        iso
    }));

const LanguageSelectorComp: React.SFC<LanguageSelectorProps> = (props): JSX.Element => {
    const options = composeLanguageMap(props.languages);
    return (
        <div className={`${languageSelector} select-small`}>
            <Select
                data-spec="language-selector"
                // Flow
                value={props.iso}
                onChange={props.onChange}
                // LanguageSelector
                valueKey="iso"
                labelKey="name"
                searchable={false}
                clearable={false}
                options={options}
            />
        </div>
    );
};

export default LanguageSelectorComp;
