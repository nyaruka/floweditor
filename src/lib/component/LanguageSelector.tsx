import * as React from 'react';
import Select from 'react-select';
import { Languages, SetLanguage } from '../flowTypes';

import { languageSelector } from './LanguageSelector.scss';

export interface Language {
    name: string;
    iso: string;
}

export interface LanguageSelectorProps {
    iso: string;
    languages: Languages;
    onChange: (language: Language) => void;
}

export const composeLanguageMap = (languages: Languages): Language[] =>
    Object.keys(languages).map(iso => ({
        name: languages[iso],
        iso
    }));

const LanguageSelectorComp: React.SFC<LanguageSelectorProps> = (
    props
): JSX.Element => {
    const className = `${languageSelector} select-small`;
    const options = composeLanguageMap(props.languages);
    return (
        <div className={className}>
            <Select
                data-spec="language-selector"
                value={props.iso}
                onChange={props.onChange}
                // LanguageSelector
                valueKey="iso"
                labelKey="name"
                searchable={false}
                clearable={false}
                options={options}
            />
        </div>
    );
};

export default LanguageSelectorComp;
