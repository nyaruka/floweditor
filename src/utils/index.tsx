import * as React from 'react';
import { Language } from '../component/LanguageSelector';
import { Action, Case, Exit, Languages, LocalizationMap, ContactProperties } from '../flowTypes';
import Localization, { LocalizedObject } from '../services/Localization';
import { SearchResult } from '../store';
import * as variables from '../variables.scss';

export const V4_UUID = /^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/i;
const LABEL_CHARS = /^[a-zA-Z0-9-][a-zA-Z0-9- ]*$/;
const MAX_LABEL_LENGTH = 36;
const SNAKED_CHARS = /\s+(?=\S)/g;

export const GRID_SIZE: number = parseInt(variables.grid_size, 10) || 20;
export const NODE_SPACING: number = parseInt(variables.node_spacing, 10) || 10;
export const NODE_PADDING: number = parseInt(variables.node_padding, 10) || 10;
export const QUIET_NOTE = 2000;
export const CONFIRMATION_TIME = 2000;

interface BoolMap {
    [key: string]: boolean;
}

interface Bounds {
    left: number;
    top: number;
    right: number;
    bottom: number;
}

/**
 * Adjusts the left and top offsets to a grid
 * @param left horizontal offset
 * @param top vertical offset
 */
export const snapToGrid = (left: number, top: number): { left: number; top: number } => {
    let leftAdjust = left % GRID_SIZE;
    let topAdjust = top % GRID_SIZE;

    if (leftAdjust > GRID_SIZE / 3) {
        leftAdjust = GRID_SIZE - leftAdjust;
    } else {
        leftAdjust = leftAdjust * -1;
    }

    if (topAdjust > GRID_SIZE / 3) {
        topAdjust = GRID_SIZE - topAdjust;
    } else {
        topAdjust = topAdjust * -1;
    }

    return {
        left: Math.max(left + leftAdjust, 0),
        top: Math.max(top + topAdjust, 0)
    };
};

/**
 * Turns a string array into a bool map for constant lookup
 * @param {string[]} array - an array of strings, e.g. contact fields
 * @returns {object} A map of each string
 */
export const toBoolMap = (array: string[]): BoolMap =>
    array.reduce(
        (map: BoolMap, item: string) => ({
            ...map,
            [item]: true
        }),
        {}
    );

/**
 * Inserts commas into numbers where appropriate for better readability
 * @param {number} value - A number, e.g. 10000
 * @returns {string} A comma-separated string, e.g. 10,000
 */
export const addCommas = (value: number): string =>
    value.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');

/**
 * Replaces spaces with underscores
 * @param {string} value - A space-separated string to be snaked, e.g. a raw flow field name ('my flow field')
 * @returns {string} A snaked string, e.g. 'my_flow_field'
 */
export const snakify = (value: string): string =>
    value
        .toLowerCase()
        .trim()
        .replace(SNAKED_CHARS, '_');
/**
 * Returns true if a given UUID matches v4 format
 * @param {string} uuid - A version 4 UUID (no braces, uppercase OK)
 * @returns {boolean}
 */
export const validUUID = (uuid: string): boolean => V4_UUID.test(uuid);

/**
 * Returns a given string in title case, e.g. 'full name' becomes 'Full Name'
 * @param {string} str - string to be title-cased
 * @returns {string} Title-cased string
 */
export const titleCase = (str: string): string =>
    str.replace(/\b\w+/g, s => s.charAt(0).toUpperCase() + s.substr(1).toLowerCase());

export const getSelectClass = (errors: number): string => {
    if (errors === 0) {
        return '';
    }
    // We use a global selector here for react-select
    return 'select-invalid';
};

export const reorderList = (list: any[], startIndex: number, endIndex: number): any[] => {
    const [removed] = list.splice(startIndex, 1);

    list.splice(endIndex, 0, removed);

    return list;
};

/**
 * Compares basic objects (no methods and DOM nodes; property order important)
 * @param {object} objA - basic object
 * @param {object} objB - basic object
 * @returns {boolean}
 */
export const jsonEqual = (objA: {}, objB: {}): boolean =>
    JSON.stringify(objA) === JSON.stringify(objB);

/**
 * Checks whether any of a list of error strings contain one or more queries.
 * Used to determine whether a particular error has been encountered.
 */
export const hasErrorType = (errors: string[], exps: RegExp[]): boolean => {
    if (!errors.length) {
        return false;
    }
    for (const error of errors) {
        for (const exp of exps) {
            if (error.match(exp)) {
                return true;
            }
        }
    }
    return false;
};

export const getLocalizedObject = (localizations: LocalizedObject[]) => {
    if (localizations && localizations.length) {
        return localizations[0];
    }
};

export interface ClickHandler {
    onMouseDown: (event: React.MouseEvent<HTMLDivElement>) => void;
    onMouseUp: (event: React.MouseEvent<HTMLDivElement>) => void;
}

/**
 * Creates a simple click handler via onMouseDown and onMouseUp.
 * This is a necessity in order to let jsPlumb manage our element dragging.
 * @param onClick
 */
export const createClickHandler = (
    onClick: (event: React.MouseEvent<HTMLDivElement>) => void
): ClickHandler => {
    return {
        onMouseDown: (event: React.MouseEvent<HTMLDivElement>) => {
            this._clicked = true;
        },
        onMouseUp: (event: React.MouseEvent<HTMLDivElement>) => {
            if (this._clicked) {
                onClick(event);
            }
            this._clicked = false;
        }
    };
};

/**
 * Callback to pass to Array.prototype.map to return SearchResult[].
 * Use on 'results' property of payload returned from an endpoint returning
 * groups, fields.
 */
export const resultsToSearchOpts = ({ name, uuid, type }: any): SearchResult => ({
    name,
    id: uuid,
    type
});

/**
 * Get the first language in a Languages map
 */
export const getBaseLanguage = (languages: Languages): Language => {
    const [iso] = Object.keys(languages);
    const name = languages[iso];
    return {
        name,
        iso
    };
};

/**
 * Get a language from a Languages map in Language format
 */
export const getLanguage = (languages: Languages, iso: string): Language => ({
    name: languages[iso],
    iso
});

export const getLocalization = (
    obj: Action | Exit | Case,
    localization: LocalizationMap,
    iso: string,
    languages: Languages
) => Localization.translate(obj, iso, languages, localization[iso]);

export const dump = (thing: any) => console.log(JSON.stringify(thing, null, 2));

/**
 * Apply emphasis style
 */
export const emphasize = (text: string) => <span className="emph">{text}</span>;

/**
 * Does property exist in our ContactProperties enum?
 * @param {string} propertyToCheck - 'name' key on react-select AutoCompleteOption
 */
export const propertyExists = (propertyToCheck: string) => {
    for (const property of Object.keys(ContactProperties)) {
        if (property.toLowerCase() === propertyToCheck.toLowerCase().trim()) {
            return true;
        }
    }
    return false;
};

/**
 * Should x element be rendered?
 */
export const renderIf = (condition: boolean) => (elem: JSX.Element) => (condition ? elem : null);

/**
 * Does the label meet our length requirements?
 * @param {string} label - label created by react-select
 */
export const properLabelLength = (label: string = '') =>
    label.length > 0 && label.length <= MAX_LABEL_LENGTH;

/**
 * Does the label meet our character requirements?
 * @param {string} label
 */
export const containsOnlyLabelChars = (label: string = '') => LABEL_CHARS.test(label);

/**
 * Does the label meet our length, character requirements?
 * @param {string} label - label created by react-select
 */
export const isValidLabel = (label: string) =>
    properLabelLength(label) && containsOnlyLabelChars(label);
