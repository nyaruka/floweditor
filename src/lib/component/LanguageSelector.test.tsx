import * as React from 'react';
import { shallow } from 'enzyme';
import { getSpecWrapper } from '../utils';
import { Languages } from '../flowTypes';
import LanguageSelector, {
    LanguageSelectorProps,
    composeLanguageMap
} from './LanguageSelector';

const iso: string = 'spa';

const languages: Languages = {
    eng: 'English',
    spa: 'Spanish',
    fre: 'French'
};

const props: LanguageSelectorProps = {
    iso,
    languages,
    onChange: jest.fn()
};

describe('LanguageSelector >', () => {
    describe('helpers >', () => {
        describe('composeLanguageMap', () => {
            it("should return an array of language maps with 'name' and 'iso' keys", () => {
                expect(composeLanguageMap(props.languages)).toMatchSnapshot();
            });
        });
    });

    describe('render >', () =>
        it('should render Select component with expected props', () => {
            const wrapper = shallow(<LanguageSelector {...props} />);

            expect(
                getSpecWrapper(wrapper, 'language-selector').props()
            ).toEqual(
                expect.objectContaining({
                    value: iso,
                    onChange: props.onChange,
                    options: composeLanguageMap(languages)
                })
            );
        }));
});
