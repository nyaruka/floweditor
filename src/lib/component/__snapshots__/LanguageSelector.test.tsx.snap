// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguageSelector > helpers > composeLanguageMap should return an array of language maps with 'name' and 'iso' keys 1`] = `
Array [
  Object {
    "iso": "eng",
    "name": "English",
  },
  Object {
    "iso": "spa",
    "name": "Spanish",
  },
  Object {
    "iso": "fre",
    "name": "French",
  },
]
`;
